In [116]:
# Importing all the required packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import math
%matplotlib inline

In [117]:
# setting seed value
seed=250
np.random.seed(250)

1. Read the dataset in file1.csv into a pandas dataframe.

In [118]:
dataset=pd.read_csv('file1.csv',index_col=False)
# header = None is not used here because we need header titles to understand the description 
dataset

attr1  attr2  attr3  attr4  attr5  attr6  attr7   atr8  attr9  attr10  \
0     42.9   26.0   31.5   13.1   10.4   18.8   14.1  106.2   71.5    32.5   
1     43.7   28.5   33.5   14.0   11.8   20.6   15.1  110.5   79.0    34.4   
2     40.1   28.2   33.3   13.9   10.9   19.7   14.1  115.1   83.2    33.4   
3     44.3   29.9   34.0   13.9   11.2   20.9   15.0  104.5   77.8    31.0   
4     41.0   26.8   32.2   14.9   10.6   17.8   14.0  107.5   78.0    36.0   
..     ...    ...    ...    ...    ...    ...    ...    ...    ...     ...   
441   38.5   25.6   31.7   11.8   10.2   16.8   12.9   92.8   74.7    28.1   
442   35.3   25.4   28.5   10.6   10.4   17.0   12.6   95.9   60.8    26.8   
443   36.2   28.7   27.8   12.4   10.2   17.9   13.2   96.1   66.9    26.2   
444   34.9   28.7   32.9   13.4   10.8   17.9   12.9   98.5   68.8    29.6   
445   38.7   29.7   32.0   12.8   10.5   18.4   13.8  103.8   63.7    26.4   

     attr11  attr12  attr13  attr14  attr15  attr16  attr17  attr18  
0      26.0    34.5    36.5    23.5    16.5      21       1    65.6  
1      28.0    36.5    37.5    24.5    17.0      23       1    71.8  
2      28.8    37.0    37.3    21.9    16.9      28       1    80.7  
3      26.2    37.0    34.8    23.0    16.6      23       1    72.6  
4      29.0    34.5    35.0    22.0    16.5      23       1    71.0  
..      ...     ...     ...     ...     ...     ...     ...     ...  
441    23.1    34.4    32.7    20.9    15.2      19       0    60.0  
442    22.8    34.4    33.1    21.0    15.4      31       0    46.8  
443    22.6    35.0    35.2    22.4    15.2      34       0    57.3  
444    25.1    34.4    35.0    21.0    15.9      34       0    64.1  
445    23.0    36.5    34.1    21.3    15.0      24       0    63.6  

[446 rows x 18 columns]

2. Display the corresponding data description file desc1.txt and understand the contents of the data in file1.csv

In [119]:
file_obj=open('desc1.txt.txt','r')
print(file_obj.read())
file_obj.close()

Body measurements of 446 physically active individuals.

Description
Body girth measurements and skeletal diameter measurements, as well as age, weight and gender, are given for 446 physically active individuals - 223 men and 223 women. 

Format
A data frame with 446 observations on the following 18 variables.

attr1
A numerical vector, respondent's biacromial diameter in centimeters.

attr2
A numerical vector, respondent's biiliac diameter (pelvic breadth) in centimeters.

attr3
A numerical vector, respondent's bitrochanteric diameter in centimeters.

attr4
A numerical vector, respondent's elbow diameter in centimeters, measured as sum of two elbows.

attr5
A numerical vector, respondent's wrist diameter in centimeters, measured as sum of two wrists.

attr6
A numerical vector, respondent's knee diameter in centimeters, measured as sum of two knees.

attr7
A numerical vector, respondent's ankle diameter in centimeters, measured as sum of two ankles.

attr8
A numerical vector, responden

3. Replace the column names of data frame with meaningful column names, designed by you using the description in desc1.txt.

In [120]:
'''
Columns naming done wrong
'''
columns=['biacromial','biiliac','bitrochanteric','elbow','wrist D','knee D','ankle D','shoulder','waist','bicep','forearm','knee G','calf','ankle G','wrist G','age','Gender','Weight']
dataset.columns=columns
dataset

biacromial  biiliac  bitrochanteric  elbow  wrist D  knee D  ankle D  \
0          42.9     26.0            31.5   13.1     10.4    18.8     14.1   
1          43.7     28.5            33.5   14.0     11.8    20.6     15.1   
2          40.1     28.2            33.3   13.9     10.9    19.7     14.1   
3          44.3     29.9            34.0   13.9     11.2    20.9     15.0   
4          41.0     26.8            32.2   14.9     10.6    17.8     14.0   
..          ...      ...             ...    ...      ...     ...      ...   
441        38.5     25.6            31.7   11.8     10.2    16.8     12.9   
442        35.3     25.4            28.5   10.6     10.4    17.0     12.6   
443        36.2     28.7            27.8   12.4     10.2    17.9     13.2   
444        34.9     28.7            32.9   13.4     10.8    17.9     12.9   
445        38.7     29.7            32.0   12.8     10.5    18.4     13.8   

     shoulder  waist  bicep  forearm  knee G  calf  ankle G  wrist G  age  \
0       106.2   71.5   32.5     26.0    34.5  36.5     23.5     16.5   21   
1       110.5   79.0   34.4     28.0    36.5  37.5     24.5     17.0   23   
2       115.1   83.2   33.4     28.8    37.0  37.3     21.9     16.9   28   
3       104.5   77.8   31.0     26.2    37.0  34.8     23.0     16.6   23   
4       107.5   78.0   36.0     29.0    34.5  35.0     22.0     16.5   23   
..        ...    ...    ...      ...     ...   ...      ...      ...  ...   
441      92.8   74.7   28.1     23.1    34.4  32.7     20.9     15.2   19   
442      95.9   60.8   26.8     22.8    34.4  33.1     21.0     15.4   31   
443      96.1   66.9   26.2     22.6    35.0  35.2     22.4     15.2   34   
444      98.5   68.8   29.6     25.1    34.4  35.0     21.0     15.9   34   
445     103.8   63.7   26.4     23.0    36.5  34.1     21.3     15.0   24   

     Gender  Weight  
0         1    65.6  
1         1    71.8  
2         1    80.7  
3         1    72.6  
4         1    71.0  
..      ...     ...  
441       0    60.0  
442       0    46.8  
443       0    57.3  
444       0    64.1  
445       0    63.6  

[446 rows x 18 columns]

4. Display the maximum, minimum, median, first quartile, third quartile information for each column in the dataframe. Use an appropriate pandas command.

In [121]:
dataset.describe().T # Transposed the matrix

count        mean        std   min     25%     50%      75%  \
biacromial      446.0   38.823991   3.052158  32.6  36.200   38.70   41.175   
biiliac         446.0   27.771525   2.201475  18.7  26.500   27.85   29.200   
bitrochanteric  446.0   31.944843   2.002236  26.3  30.600   32.00   33.300   
elbow           446.0   13.352915   1.359206   9.9  12.400   13.20   14.400   
wrist D         446.0   10.534753   0.958781   8.1   9.800   10.50   11.200   
knee D          446.0   18.772646   1.352456  15.7  17.900   18.70   19.600   
ankle D         446.0   13.830493   1.256878   9.9  12.900   13.80   14.800   
shoulder        446.0  108.186323  10.501635  85.9  99.325  108.30  116.675   
waist           446.0   77.090135  11.237657  57.9  68.000   75.55   84.650   
bicep           446.0   31.041928   4.249316  22.4  27.400   30.85   34.400   
forearm         446.0   25.898206   2.836281  19.6  23.400   25.80   28.375   
knee G          446.0   36.161435   2.632202  29.0  34.400   36.00   37.775   
calf            446.0   36.034529   2.893242  28.4  34.000   35.90   37.900   
ankle G         446.0   22.131390   1.894831  16.4  20.925   22.00   23.275   
wrist G         446.0   16.098430   1.401595  13.0  15.000   16.10   17.100   
age             446.0   30.340807   9.644256  18.0  23.000   28.00   36.000   
Gender          446.0    0.500000   0.500561   0.0   0.000    0.50    1.000   
Weight          446.0   69.006054  13.474136  42.0  58.325   67.95   78.750   

                  max  
biacromial       46.2  
biiliac          34.7  
bitrochanteric   37.8  
elbow            16.7  
wrist D          13.3  
knee D           24.3  
ankle D          17.2  
shoulder        134.8  
waist           113.2  
bicep            42.3  
forearm          32.5  
knee G           49.0  
calf             47.7  
ankle G          29.3  
wrist G          19.6  
age              65.0  
Gender            1.0  
Weight          116.4

5. Use an appropriate pandas command to check if any column in the dataframe contains any missing value or not. Drop those rows if there are missing values in the row

In [122]:
# pandas command to check for number of NULL values in each column
dataset.isnull().sum()

biacromial        0
biiliac           0
bitrochanteric    0
elbow             0
wrist D           0
knee D            0
ankle D           0
shoulder          0
waist             0
bicep             0
forearm           0
knee G            0
calf              0
ankle G           0
wrist G           0
age               0
Gender            0
Weight            0
dtype: int64

In [123]:
# Total number of null values in the entire dataset
dataset.isnull().sum().sum()

0

- Since there are no NULL values in the dataset so I cannot remove any Rows. However I will be writing the code for removing rows wiyh NULL values.

In [124]:
'''
Demo code for dropping rows

dataset.dropna(inplace=True)

'''

'\nDemo code for dropping rows\n\ndataset.dropna(inplace=True)\n\n'

1. Consider the weight attribute as the response variable and all other attributes as predictor variables
2. So Y=Weight            ||    beta=Columns from 0 to 17

- Perform standardization of each column in the data frame and create a new data frame.

In [125]:
# Standardization
# Storing response variable mean
y_mean=dataset['Weight'].mean()
y_std=dataset['Weight'].std()
new_dataset=dataset

# Standardization

for i in columns:
    #if(i=='Weight'):
        #break
    #print(dataset[i].mean())
    new_dataset[i] = (dataset[i] - dataset[i].mean()) / dataset[i].std()
new_dataset

biacromial   biiliac  bitrochanteric     elbow   wrist D    knee D  \
0      1.335451 -0.804699       -0.222173 -0.186075 -0.140547  0.020226   
1      1.597561  0.330903        0.776710  0.476076  1.319641  1.351138   
2      0.418068  0.194631        0.676822  0.402504  0.380949  0.685682   
3      1.794143  0.966841        1.026431  0.402504  0.693846  1.572956   
4      0.712941 -0.441306        0.127436  1.138227  0.068052 -0.719170   
..          ...       ...             ...       ...       ...       ...   
441   -0.106151 -0.986395       -0.122285 -1.142516 -0.349145 -1.458565   
442   -1.154590 -1.077244       -1.720498 -2.025385 -0.140547 -1.310686   
443   -0.859717  0.421752       -2.070107 -0.701082 -0.349145 -0.645230   
444   -1.285645  0.421752        0.477045  0.034642  0.276650 -0.645230   
445   -0.040624  0.875993        0.027548 -0.406792 -0.036247 -0.275533   

      ankle D  shoulder     waist     bicep   forearm    knee G      calf  \
0    0.214426 -0.189144 -0.497447  0.343131  0.035890 -0.631196  0.160882   
1    1.010048  0.220316  0.169952  0.790262  0.741039  0.128624  0.506515   
2    0.214426  0.658343  0.543696  0.554930  1.023098  0.318579  0.437389   
3    0.930486 -0.351024  0.063168 -0.009867  0.106405  0.318579 -0.426694   
4    0.134863 -0.065354  0.080966  1.166793  1.093613 -0.631196 -0.357567   
..        ...       ...       ...       ...       ...       ...       ...   
441 -0.740321 -1.465136 -0.212690 -0.692330 -0.986576 -0.669187 -1.152523   
442 -0.979008 -1.169944 -1.449602 -0.998261 -1.092348 -0.669187 -1.014270   
443 -0.501634 -1.150899 -0.906785 -1.139461 -1.162863 -0.441241 -0.288441   
444 -0.740321 -0.922363 -0.737710 -0.339332 -0.281427 -0.669187 -0.357567   
445 -0.024261 -0.417680 -1.191541 -1.092394 -1.021833  0.128624 -0.668637   

      ankle G   wrist G       age    Gender    Weight  
0    0.722286  0.286509 -0.968536  0.998878 -0.252785  
1    1.250038  0.643245 -0.761158  0.998878  0.207356  
2   -0.122117  0.571898 -0.242715  0.998878  0.867881  
3    0.458410  0.357856 -0.761158  0.998878  0.266729  
4   -0.069341  0.286509 -0.761158  0.998878  0.147983  
..        ...       ...       ...       ...       ...  
441 -0.649868 -0.641006 -1.175913 -0.998878 -0.668396  
442 -0.597093 -0.498311  0.068351 -0.998878 -1.648050  
443  0.141759 -0.641006  0.379417 -0.998878 -0.868780  
444 -0.597093 -0.141575  0.379417 -0.998878 -0.364109  
445 -0.438768 -0.783700 -0.657470 -0.998878 -0.401217  

[446 rows x 18 columns]

- Display the maximum, minimum, median, first quartile, third quartile information for each column in the dataframe. Use an appropriate pandas command.

In [126]:
new_dataset.describe().T # Transposed the matrix

count          mean  std       min       25%       50%  \
biacromial      446.0 -5.098064e-16  1.0 -2.039210 -0.859717 -0.040624   
biiliac         446.0  1.449762e-15  1.0 -4.120658 -0.577579  0.035647   
bitrochanteric  446.0 -1.354173e-15  1.0 -2.819269 -0.671670  0.027548   
elbow           446.0  1.274516e-15  1.0 -2.540391 -0.701082 -0.112503   
wrist D         446.0  5.735323e-16  1.0 -2.539425 -0.766341 -0.036247   
knee D          446.0  4.779435e-16  1.0 -2.271900 -0.645230 -0.053714   
ankle D         446.0  2.549032e-16  1.0 -3.127188 -0.740321 -0.024261   
shoulder        446.0  8.284355e-16  1.0 -2.122177 -0.843804  0.010825   
waist           446.0 -6.372581e-16  1.0 -1.707663 -0.808899 -0.137051   
bicep           446.0 -5.098064e-16  1.0 -2.033722 -0.857062 -0.045167   
forearm         446.0  1.656871e-15  1.0 -2.220586 -0.880803 -0.034625   
knee G          446.0  8.921613e-16  1.0 -2.720701 -0.669187 -0.061331   
calf            446.0  2.357855e-15  1.0 -2.638745 -0.703200 -0.046498   
ankle G         446.0  1.688734e-15  1.0 -3.024751 -0.636674 -0.069341   
wrist G         446.0 -5.098064e-16  1.0 -2.210646 -0.783700  0.001120   
age             446.0  0.000000e+00  1.0 -1.279602 -0.761158 -0.242715   
Gender          446.0  1.274516e-16  1.0 -0.998878 -0.998878  0.000000   
Weight          446.0 -4.460806e-16  1.0 -2.004288 -0.792708 -0.078376   

                     75%       max  
biacromial      0.770278  2.416653  
biiliac         0.648872  3.147197  
bitrochanteric  0.676822  2.924308  
elbow           0.770365  2.462530  
wrist D         0.693846  2.884127  
knee D          0.611742  4.086901  
ankle D         0.771361  2.680854  
shoulder        0.808320  2.534241  
waist           0.672726  3.213291  
bicep           0.790262  2.649384  
forearm         0.873254  2.327623  
knee G          0.613010  4.877500  
calf            0.644768  4.031972  
ankle G         0.603542  3.783246  
wrist G         0.714593  2.498274  
age             0.586794  3.593765  
Gender          0.998878  0.998878  
Weight          0.723159  3.517401

6. Split the data into two sets such that 80% of the data is considered as set T1 and 20% of the data is considered as set T2.

In [127]:
T1_set=new_dataset.sample(frac=0.8,random_state=seed)  # 250 is used to maintain seed value
T2_set=new_dataset.drop(T1_set.index)
print(len(T1_set),len(T2_set))

357 89


7.  Using T1 as training data, solve an ordinary least squares regression problem and obtain the regression coefficients. Using these coefficients, compute the R2 and mean squared error values on the training set T1 and test set T2

In [128]:
response_cols=['Weight']
predictor_columns=list(dataset.drop(columns=['Weight']).columns)  # Extracting the predictor columns
#print(predictor_columns,len(predictor_columns))
T1_train=T1_set.drop(columns=response_cols,axis=1) 
T2_test=T2_set.drop(columns=response_cols,axis=1) 
T1_train
#predictor_columns

biacromial   biiliac  bitrochanteric     elbow   wrist D    knee D  \
445   -0.040624  0.875993        0.027548 -0.406792 -0.036247 -0.275533   
176    1.007814  0.012935        0.676822  1.285372  0.798145  0.315984   
76     0.647414 -0.305034       -1.271000 -0.038931 -0.140547 -0.201593   
35    -0.171679 -1.168092       -0.871447 -0.112503 -0.140547  0.020226   
249    0.090431 -0.123338        1.026431 -0.553937 -0.557743 -0.571291   
..          ...       ...             ...       ...       ...       ...   
92    -0.401025 -1.531485       -1.820386 -0.333220  0.276650 -0.571291   
252   -1.252881 -0.804699       -1.071224 -0.701082 -1.079238 -1.162807   
208    0.680177 -1.531485       -0.671670  0.328931  1.006743  0.168105   
12     1.138869 -0.623003       -0.372006  0.917510  0.276650 -0.127653   
290   -0.204443  1.012265        1.775593 -0.553937  0.276650  0.537802   

      ankle D  shoulder     waist     bicep   forearm    knee G      calf  \
445 -0.024261 -0.417680 -1.191541 -1.092394 -1.021833  0.128624 -0.668637   
176  0.850923  1.001147  0.285635  1.096193  1.164128  0.850453  0.333699   
76  -0.819883  0.001302 -0.052514  0.484330  0.458979 -0.251286 -0.253878   
35  -0.660759 -0.665261 -0.639825  0.107799  0.247434 -1.277043  0.091755   
249 -0.899446 -1.055676 -0.630926 -1.115927 -1.162863  0.394561 -0.011934   
..        ...       ...       ...       ...       ...       ...       ...   
92   0.373550  0.087003 -0.185994  1.237392  1.058355 -0.631196 -0.876017   
252 -1.217694 -1.807940 -1.431805 -1.539525 -1.198120 -1.391016 -1.048833   
208  1.010048  0.839267  0.588189  0.343131  0.423722 -0.289277  0.333699   
12  -0.024261  0.905923 -0.230487  0.037199  0.353207  0.014651 -0.945144   
290 -0.183386 -0.512903  0.525898 -0.292265 -0.457714  0.736480  0.610205   

      ankle G   wrist G       age    Gender  
445 -0.438768 -0.783700 -0.657470 -0.998878  
176  1.830565  1.285371 -0.968536  0.998878  
76  -0.333217  0.072467 -1.072224  0.998878  
35  -1.124845 -0.212922 -1.072224  0.998878  
249 -0.597093 -0.783700 -0.761158 -0.998878  
..        ...       ...       ...       ...  
92   0.458410  0.714593 -0.553781  0.998878  
252 -0.860969 -1.140437 -1.175913 -0.998878  
208  0.300085  0.357856  0.897860  0.998878  
12   0.722286 -0.070227 -0.864847  0.998878  
290 -0.755418  0.357856  1.727369 -0.998878  

[357 rows x 17 columns]

In [129]:
T2_test

biacromial   biiliac  bitrochanteric     elbow   wrist D    knee D  \
2      0.418068  0.194631        0.676822  0.402504  0.380949  0.685682   
3      1.794143  0.966841        1.026431  0.402504  0.693846  1.572956   
4      0.712941 -0.441306        0.127436  1.138227  0.068052 -0.719170   
9      1.564797  0.694296        1.226207  1.506089  0.380949  1.425077   
14     1.957962  0.012935        0.127436  1.358944  1.111042  0.463863   
..          ...       ...             ...       ...       ...       ...   
419   -0.466552  0.558024        0.477045 -0.480365 -0.557743 -0.423412   
429   -1.580518 -0.850123       -1.271000 -1.436806 -0.974939 -1.606444   
432   -0.761425 -0.123338       -0.571782 -1.289661 -0.974939 -1.310686   
436    0.188722  0.240055       -0.871447 -0.406792 -0.036247  0.537802   
438   -0.990771 -0.214186       -1.720498 -1.583950 -0.349145 -0.940989   

      ankle D  shoulder     waist     bicep   forearm    knee G      calf  \
2    0.214426  0.658343  0.543696  0.554930  1.023098  0.318579  0.437389   
3    0.930486 -0.351024  0.063168 -0.009867  0.106405  0.318579 -0.426694   
4    0.134863 -0.065354  0.080966  1.166793  1.093613 -0.631196 -0.357567   
9    1.328297  0.982102  0.908540  1.402125  1.622474  1.382327  2.061864   
14   0.691799  1.182071  0.383520  1.966922  2.116079  0.888444  1.785357   
..        ...       ...       ...       ...       ...       ...       ...   
419 -0.024261 -0.950930 -0.995771 -1.045328 -0.775031 -0.213295 -0.703200   
429 -0.979008 -1.255645 -0.862291 -0.409931 -0.986576 -0.213295  0.471952   
432 -1.695068 -1.436569 -0.675420 -1.563058 -1.339150 -0.859142 -1.809226   
436  0.453112 -0.093921 -0.657622 -0.268732 -0.316684  0.584516  0.575642   
438 -0.740321 -1.503225 -1.120352 -1.304193 -1.021833 -0.631196 -1.152523   

      ankle G   wrist G       age    Gender  
2   -0.122117  0.571898 -0.242715  0.998878  
3    0.458410  0.357856 -0.761158  0.998878  
4   -0.069341  0.286509 -0.761158  0.998878  
9    2.305541  1.927496 -0.864847  0.998878  
14   0.352860  1.499412 -0.450092  0.998878  
..        ...       ...       ...       ...  
419 -0.227667 -0.712353 -0.761158 -0.998878  
429  0.194534 -1.211784  0.794171 -0.998878  
432 -1.177620 -1.425826  0.690483 -0.998878  
436  0.827836  0.072467 -1.175913 -0.998878  
438 -0.966519 -0.855047 -0.450092 -0.998878  

[89 rows x 17 columns]

In [130]:
# converting predictor variables to numpy array
T1_train=T1_train.to_numpy()
T2_test=T2_test.to_numpy()
print(T1_train.shape,T2_test.shape)

(357, 17) (89, 17)


In [131]:
# Add a column of 1's to account for the beta0 in the matrix multiplication Y=beta0 + beta1*x1+....
X_train=np.hstack((T1_train, np.ones((T1_train.shape[0], 1), dtype=T1_train.dtype)))
X_test=np.hstack((T2_test, np.ones((T2_test.shape[0], 1), dtype=T2_test.dtype)))
print(X_train.shape,X_test.shape)

(357, 18) (89, 18)


In [132]:
# Convert respective response variables into numpy arrays
Y_train=T1_set[response_cols].to_numpy()
Y_test=T2_set[response_cols].to_numpy()
print(Y_train.shape,Y_test.shape)

(357, 1) (89, 1)


10. To perform ℓ2 regularized least squares regression, choose regularization constant λ ∈ {10−5, 10−4, 10−3, 10−2, 10−1, 1, 10, 100, 1000, 10000} using a 5-fold cross-validation on the full train set T1. Use average R2 scores of the folds as the criterion to choose best regularization parameter λ⋆. Perform ℓ2 regularized least squares regression with best regularization parameter λ ⋆ obtained from the cross-validation procedure on the set T1 and obtain the regression coefficients.

In [133]:
# Function to compute R2 and MSE

def compute_Rsquared_and_MSE(beta, X, Y): 

  #first we shall compute y_bar 
  y_bar = Y.mean()
  #print('y_bar',y_bar)

  sigma_yy = 0
  for i in range(Y.shape[0]): 
       y_i = Y[i] 
       sigma_yy += (y_i-y_bar)**2
  #print('sigma_yy:', sigma_yy)

  #print('\n_________________________\n')

  #now we can compute sum of squared residuals 
  sum_sq_residuals = 0.
  for i in range(Y.shape[0]): #access each row from the data set 
    x_i = X[i,:] #access i-th row of X
    y_i = Y[i] #access i-th row of y

    y_pred_i = np.dot(x_i,beta) #compute the prediction obtained using the regression coefficients 
    e_i = y_i - y_pred_i #compute the difference between the actual observation y_i and prediction y_pred_i

    sum_sq_residuals += (e_i)**2

    
  #Then we will compute the R^2 quantity and MSE
  R_sq = 1-sum_sq_residuals/sigma_yy
  MSE = sum_sq_residuals/Y.shape[0]
  #print('R^2 :',R_sq, ' MSE:',MSE)
  return R_sq, MSE


In [134]:
import scipy
from scipy import linalg

#print(T1_set.shape)

seeds = [100, 200,300,400,500] # Since 5 fold cross validation is chosen so 5 seeds are given

lambda_val = [pow(10,-5),pow(10,-4),pow(10,-3),pow(10,-2),pow(10,-1),1,10,100,1000,10000]

train_R2_seeds = [] 
val_R2_seeds = [] 

train_MSE_seeds = [] 
val_MSE_seeds = [] 
for seed in seeds: 
  T1_train = T1_set.sample(frac=0.8,random_state=seed) # since 5 five fold cross validation so dataset is divided into 5 chunks. So validation frac=0.2
  T1_val = T1_set.drop(T1_train.index)
  #print(T1_train.shape,T1_val.shape)
  #print(T1_train.Weight)
  X_val = T1_val.drop(columns=response_cols,axis=1)
  X_train = T1_train.drop(columns=response_cols,axis=1)
  

  X_train = X_train.to_numpy()
  X_val = X_val.to_numpy()

  X_train = np.hstack((X_train, np.ones((X_train.shape[0], 1), dtype=X_train.dtype)))
  X_val = np.hstack((X_val, np.ones((X_val.shape[0], 1), dtype=X_val.dtype)))
  
  Y_train = T1_train[response_cols].to_numpy()
  Y_val = T1_val[response_cols].to_numpy()

  train_R2_lambdas = []
  val_R2_lambdas = []

  train_MSE_lambdas = []
  val_MSE_lambdas = []
  
  for lambda_ in lambda_val:
    XTX = np.matmul(np.transpose(X_train),X_train)
    Xy = np.matmul(np.transpose(X_train),Y_train)
    I = np.identity(X_train.shape[1])

    beta =scipy.linalg.solve( XTX+lambda_*I,Xy)
    #print('lambda:',lambda_)
    #print('beta:',beta)
    #print('***************')

    train_R2, train_MSE = compute_Rsquared_and_MSE(beta, X_train, Y_train)
    train_R2_lambdas.append(train_R2)
    train_MSE_lambdas.append(train_MSE)

    val_R2, val_MSE = compute_Rsquared_and_MSE(beta, X_val, Y_val)
    val_R2_lambdas.append(val_R2)
    val_MSE_lambdas.append(val_MSE)

  train_R2_seeds.append(np.array(train_R2_lambdas))
  train_MSE_seeds.append(np.array(train_MSE_lambdas))

  val_R2_seeds.append(np.array(val_R2_lambdas))
  val_MSE_seeds.append(np.array(val_MSE_lambdas))

val_R2_seeds = np.array(val_R2_seeds).squeeze()
print(val_R2_seeds)

mean_val_R2_lambdas = np.mean(val_R2_seeds,axis=0)

train_R2_seeds = np.array(train_R2_seeds).squeeze()
print(train_R2_seeds)

mean_train_R2_lambdas = np.mean(train_R2_seeds,axis=0)

print('5-fold train R2 values:', train_R2_seeds,'\n')

print('5-fold train MSEs:', train_MSE_seeds,'\n')

print('5-fold val R2 values:', mean_val_R2_lambdas.squeeze(),'\n')
print('5-fold train R2 values:', mean_train_R2_lambdas.squeeze(),'\n')

#print('k-fold val MSEs:', val_MSE_seeds)


best_lambda_idx = np.argmax(mean_val_R2_lambdas) # getting the index of lambda val with maximum 
best_lambda = lambda_val[best_lambda_idx]

print('best lambda:', best_lambda)



[[0.95494705 0.95494705 0.95494709 0.95494749 0.95495144 0.95498348
  0.95478246 0.94372225 0.84864438 0.36612563]
 [0.95321971 0.95321974 0.95322006 0.95322321 0.95325437 0.9535348
  0.95441641 0.93900612 0.82600523 0.32673116]
 [0.94328924 0.94328926 0.94328945 0.94329132 0.94330985 0.94348099
  0.94418573 0.93305495 0.84206777 0.35886553]
 [0.94713585 0.94713583 0.9471357  0.94713438 0.947121   0.94697442
  0.94474325 0.92213026 0.84039815 0.35157152]
 [0.95256805 0.95256809 0.95256846 0.95257214 0.9526087  0.95294769
  0.95468115 0.94551417 0.85049097 0.3581794 ]]
[[0.9548508  0.9548508  0.9548508  0.9548508  0.95485069 0.95484076
  0.95412199 0.93878552 0.84763939 0.37516596]
 [0.95488982 0.95488982 0.95488982 0.95488982 0.95488967 0.95487597
  0.95397548 0.93769404 0.84705181 0.37185627]
 [0.95805778 0.95805778 0.95805778 0.95805778 0.95805766 0.9580463
  0.95723769 0.94116277 0.84990966 0.37201323]
 [0.95632148 0.95632148 0.95632148 0.95632148 0.95632137 0.9563114
  0.95561958 0

In [135]:
# Obtaining beta fron the best lamda

X_train = T1_set.drop(columns=response_cols)
X_train = X_train.to_numpy()

X_train = np.hstack((X_train, np.ones((X_train.shape[0], 1), dtype=X_train.dtype)))
#print(X_train)
Y_train = T1_set[response_cols].to_numpy()

XTX = np.matmul(np.transpose(X_train),X_train)
Xy = np.matmul(np.transpose(X_train),Y_train)
I = np.identity(X_train.shape[1])

final_beta =scipy.linalg.solve( XTX+best_lambda*I,Xy)
print('lambda:',best_lambda)
print('Final beta: \n',final_beta.shape,'\n',final_beta)
#print('***************')

lambda: 10
Final beta: 
 (18, 1) 
 [[ 0.0377289 ]
 [ 0.0531302 ]
 [ 0.06092999]
 [ 0.04667624]
 [ 0.0168943 ]
 [ 0.01649402]
 [ 0.05936621]
 [ 0.16382189]
 [ 0.37934054]
 [ 0.10962631]
 [ 0.05508439]
 [ 0.15783918]
 [ 0.07660056]
 [-0.00546107]
 [-0.00348104]
 [-0.06829123]
 [-0.06943875]
 [-0.00270072]]


In [136]:
X_train=T1_set.drop(columns=response_cols).to_numpy()
X_train = np.hstack((X_train, np.ones((X_train.shape[0], 1), dtype=X_train.dtype)))
Y_train=T1_set[response_cols].to_numpy()
X_test=T2_set.drop(columns=response_cols).to_numpy()
X_test = np.hstack((X_test, np.ones((X_test.shape[0], 1), dtype=X_train.dtype)))
Y_test=T2_set[response_cols].to_numpy()

train_R2, train_MSE = compute_Rsquared_and_MSE(final_beta, X_train, Y_train)
test_R2, test_MSE = compute_Rsquared_and_MSE(final_beta, X_test, Y_test)

print('Train R^2 :', train_R2, ' Train MSE:', train_MSE)
print('Test R^2 :', test_R2, ' Test MSE:', test_MSE)

Train R^2 : [0.95529334]  Train MSE: [0.04584052]
Test R^2 : [0.94305357]  Test MSE: [0.05041673]


- Train R^2 : [0.95529334]  Train MSE: [0.04584052]
- Test R^2 : [0.94305357]  Test MSE: [0.05041673]

- From the regression coefficients obtained, indicate the relative importance of each attribute in predicting the response variable using the magnitude and sign of the regression coefficients. Compare the regression coefficients with those obtained in Question 1, discuss your observations.

In [137]:
c=0
prev_beta=[1.66673598e-01, 2.77294652e-01, 3.67516154e-01, 5.52994571e-01,1.78626675e-01,1.26196476e-01,7.62232067e-01,2.06232019e-01,5.09818572e-01,3.42720762e-01,2.11292881e-01,8.23451040e-01,3.48753766e-01,-1.33417198e-01,-4.82312890e-02,-1.11730754e-01,-2.33755497e+00,-9.08923467e+01]

for i in columns:
    if(i=='Weight'):
        break
    print(i,' : ','Beta value in current Qs :',final_beta[c],'Beta value in previous Qs :', prev_beta[c])
    c+=1

print(" \nBias term in in this Qs :",final_beta[-1],"Bias term in previous qs :",prev_beta[-1])

biacromial  :  Beta value in current Qs : [0.0377289] Beta value in previous Qs : 0.166673598
biiliac  :  Beta value in current Qs : [0.0531302] Beta value in previous Qs : 0.277294652
bitrochanteric  :  Beta value in current Qs : [0.06092999] Beta value in previous Qs : 0.367516154
elbow  :  Beta value in current Qs : [0.04667624] Beta value in previous Qs : 0.552994571
wrist D  :  Beta value in current Qs : [0.0168943] Beta value in previous Qs : 0.178626675
knee D  :  Beta value in current Qs : [0.01649402] Beta value in previous Qs : 0.126196476
ankle D  :  Beta value in current Qs : [0.05936621] Beta value in previous Qs : 0.762232067
shoulder  :  Beta value in current Qs : [0.16382189] Beta value in previous Qs : 0.206232019
waist  :  Beta value in current Qs : [0.37934054] Beta value in previous Qs : 0.509818572
bicep  :  Beta value in current Qs : [0.10962631] Beta value in previous Qs : 0.342720762
forearm  :  Beta value in current Qs : [0.05508439] Beta value in previous Qs :

## Observations:

- From comapring the beta values of the previous question and the current question we observe that the sign of the both the values remains the same which means the correlation with the response variable is the same.

- From the beta values of the current qs we can directly explain the percentage of the variance explained by the attribute since the data is standardized.

- In the current Qs beta values we have observed that the range of the magnitude of beta values is between [0,1] whereas in the previous qs beta values it exceeds 1. 

- In the current qs for beta values we can say for sure that the attributes with higher magnitude of beta values are more importnat or with greater priority than those with lower magnitude. This is because the entire dataset has been completely standardized. So the beta values in this case are directly a multiplier of the standard deviation.

- However for the beta values obtained from Qs 1 since the data is not standardized we cannot clearly comment on the importance of that attribute directly based on the magnitude of the coefficient.

- Also if we compare the bias term we can see that only a small part of the variance id explained by the current qs bias while the bias term of the prev qs is a very large value.

- Rank the attributes in the data set (excluding the bias term) based on non-increasing order of magnitude of the regression coefficients and print them in that order.

In [138]:
import operator
b=final_beta[0:17,:] # Excluding the bias term
original_dct={}
#print(b)
new_b=np.absolute(b) # converting to magnitude only
dct={}
c=0
for i in predictor_columns:
    dct[i]=float(new_b[c])
    original_dct[i]=float(b[c])
    c+=1
sorted_d = dict( sorted(dct.items(), key=operator.itemgetter(1),reverse=True))
#print(original_dct)
for i in sorted_d.keys():
    print(i,':',sorted_d[i])

waist : 0.3793405409560516
shoulder : 0.16382188651525595
knee G : 0.15783917977485865
bicep : 0.10962630905849906
calf : 0.0766005639784037
Gender : 0.06943874951993631
age : 0.06829122515469975
bitrochanteric : 0.06092998585292981
ankle D : 0.05936621035492274
forearm : 0.05508439388796437
biiliac : 0.05313019923143515
elbow : 0.04667623807012449
biacromial : 0.037728895997058205
wrist D : 0.01689429950239106
knee D : 0.016494016833018765
ankle G : 0.005461073971511808
wrist G : 0.0034810414253886378


- Consider the top 5 attributes in the previous question as d1, d2, d3, d4, d5. Perform ℓ2 regularized least squares regression with best regularization parameter λ ⋆ on the set T1 containing only attribute d1 and response variable. Compute R2 values on the set T1 containing only attribute d1 and response variable.

In [166]:
# Storing top 7 attributes.

# Here I am choosing top 7 attributes instead of 5 to display R2 and adjusted R2 values for better clarity and better conclusion.

top=[] 
c=0
for i in sorted_d.keys():
    if(c==7):
        break
    top.append(i)
    c+=1
print(top)

['waist', 'shoulder', 'knee G', 'bicep', 'calf', 'Gender', 'age']


In [167]:
def adjusted(r2,n,k):
    return ( 1-(1-r2)*(n-1)/(n-k-1))

In [168]:
temp=[] # columns to be worked upon 
Y_train=T1_set['Weight'].to_numpy()
Y_test=T2_set['Weight'].to_numpy()
#print(T1_set)
#print(Y_train)
store_r2_train=[]
store_adjusted_train=[]
store_r2_test=[]
store_adjusted_test=[]
for i in top:
    temp.append(i) # To store the column names to be used in dataset

    # Preparing dataset
    X_train=T1_set[temp].to_numpy()
    X_test=T2_set[temp].to_numpy()
    X_train=np.hstack((X_train, np.ones((X_train.shape[0], 1), dtype=X_train.dtype)))
    X_test=np.hstack((X_test, np.ones((X_test.shape[0], 1), dtype=X_test.dtype)))
    #print(X_train)

    # Finding betas
    XTX = np.matmul(np.transpose(X_train),X_train)
    Xy = np.matmul(np.transpose(X_train),Y_train)
    I = np.identity(X_train.shape[1])

    final_beta =scipy.linalg.solve( XTX+best_lambda*I,Xy)
    print('lambda:',best_lambda)
    print('Final beta:',final_beta.shape,'\n',final_beta)
    #print('***************')
    R2,MSE=compute_Rsquared_and_MSE(final_beta, X_train, Y_train)
    R2_test,MSE_test=compute_Rsquared_and_MSE(final_beta, X_test, Y_test)
    print('R2 train:',R2)
    print("Adjusted R2 train : ", adjusted(float(R2),float(Y_train.shape[0]),len(temp)))

    store_r2_train.append(R2)
    store_adjusted_train.append(adjusted(float(R2),float(Y_train.shape[0]),len(temp)))

    print('R2 test:',R2_test)
    print("Adjusted R2 test : ", adjusted(float(R2_test),float(Y_test.shape[0]),len(temp)))
    
    store_r2_test.append(R2_test)
    store_adjusted_test.append(adjusted(float(R2_test),float(Y_test.shape[0]),len(temp)))

lambda: 10
Final beta: (2,) 
 [ 0.89284235 -0.00343644]
R2 train: 0.8258034887306618
Adjusted R2 train :  0.82531279433272
R2 test: 0.795155095662152
Adjusted R2 test :  0.7928005565318319
lambda: 10
Final beta: (3,) 
 [ 0.55731419  0.41121232 -0.01048188]
R2 train: 0.8778355800448716
Adjusted R2 train :  0.8771453855818483
R2 test: 0.8659983769934034
Adjusted R2 test :  0.8628820601792966
lambda: 10
Final beta: (4,) 
 [ 0.4202136   0.3252376   0.30706127 -0.00722241]
R2 train: 0.9295552796138689
Adjusted R2 train :  0.9289565992706439
R2 test: 0.9044246781880516
Adjusted R2 test :  0.9010514315358652
lambda: 10
Final beta: (5,) 
 [ 0.37498128  0.20343466  0.29663864  0.18929594 -0.00635165]
R2 train: 0.9350235546487098
Adjusted R2 train :  0.934285185951536
R2 test: 0.913709664731167
Adjusted R2 test :  0.9096006011469369
lambda: 10
Final beta: (6,) 
 [ 0.36629089  0.19868783  0.228019    0.17075402  0.11424415 -0.00525341]
R2 train: 0.9386798476252005
Adjusted R2 train :  0.937806341

In [169]:
# printing Table
print("No. of features chosen",'\t\t',"Train_R2",'\t',"Train_Adjusted_R2",'\t',"Test_R2",'\t\t',"Test_Adjusted_R2")
for i in range(len(store_r2_train)):
    print(i+1,'\t\t\t',store_r2_train[i],'\t',store_adjusted_train[i],'\t',store_r2_test[i],'\t',store_adjusted_test[i])

No. of features chosen 		 Train_R2 	 Train_Adjusted_R2 	 Test_R2 		 Test_Adjusted_R2
1 			 0.8258034887306618 	 0.82531279433272 	 0.795155095662152 	 0.7928005565318319
2 			 0.8778355800448716 	 0.8771453855818483 	 0.8659983769934034 	 0.8628820601792966
3 			 0.9295552796138689 	 0.9289565992706439 	 0.9044246781880516 	 0.9010514315358652
4 			 0.9350235546487098 	 0.934285185951536 	 0.913709664731167 	 0.9096006011469369
5 			 0.9386798476252005 	 0.937806341181115 	 0.9190552941917435 	 0.9141791070948606
6 			 0.9387259545999276 	 0.9376755423930692 	 0.9193889232279432 	 0.913490551756817
7 			 0.94051716233196 	 0.9393240968199936 	 0.9214424395569395 	 0.9146535145803787


## Observations:

- The adjusted R2 value is always less than than the R2 value in all the cases.

- The R2 value always increases when a new feature is added to the dataset. However, this cannot be said for adjusted R2 value.

- If we look at when the 6th feature is added we can see that the adjusted R2 value decreases for both test and train sets from its previous value. This can be attributed to the fact that adjusted R2 value penalises addition of less important feature in the dataset for prediction.

- Thus we can easily conclude that adjusted R2 is a better measure for considering a feature importance to a dataset than R2 value. If large amount of unnecessary feature is added to the dataset R2 value will increase but adjusted R2 value will decrease which will represnt the actual condition of the beta values and the dataset.